File: python/131. Palindrome Partitioning.py
from typing import List

class Solution:
    def partition(self, s: str) -> List[List[str]]:
        s = list(s)
        ret = []
        self.traceback(s, ret, [], len(s))
        return ret

    def traceback(self, s, ret, tmpList, all_length):
        length = sum([len(palindrome) for palindrome in tmpList])
        if length == all_length:
            ret.append(tmpList[:])
            return

        for i in range(len(s)):
            if self.isPalindrome(s[:i+1]):
                tmpList.append("".join(s[:i+1]))
                self.traceback(s[i+1:], ret, tmpList, all_length)
                tmpList.pop()

    def isPalindrome(self, s):
        if len(s) == 0:
            return False
        low, high = 0, len(s)-1
        while(low < high):
            if s[low] != s[high]:
                return False
            low += 1
            high -= 1
        return True


if __name__ == "__main__":
    print(Solution().partition("aa"))

File: cpp/131.Palindrome Partitioning.cpp
#include <vector>
#include <string>
using namespace std;


class Solution {
private:
    vector<vector<string>> ret;

    bool isPalindrome(string s){
        int low = 0;
        int high = s.length() - 1;
        while(low < high){
            if (s[low] != s[high])
                return false;
            ++low;
            --high;
        }
        return true;
    }
    void trace_back(string s, vector<string> tmpVector, int length){
        int sum_length = 0;
        for (auto str: tmpVector){
            sum_length += str.length();
        }
        if (sum_length == length){
            vector<string>vec(tmpVector);
            ret.push_back(vec);
        }
        for (int i=0; i<s.length(); i++){
            string s1;
            s1.assign(s, 0, i+1);
            if (isPalindrome(s1)){
                tmpVector.push_back(s1);
                string s2;
                s2.assign(s, i+1, s.length()-(i+1));
                trace_back(s2, tmpVector, length);
                tmpVector.pop_back();
            }
        }
    }


public:
    vector<vector<string>> partition(string s) {
        vector<string>tmpVector;
        trace_back(s, tmpVector, s.length());
        return ret;
    }
};
